Plugins/Tweaks/Tweaks/FixScrollLearningBug.cpp

```diff
@@ -0,0 +1,146 @@
+#include "Tweaks/FixScrollLearningBug.hpp"
+
+#include "Services/Hooks/Hooks.hpp"
+#include "Utils.hpp"
+
+#include "API/CItemRepository.hpp"
+#include "API/CNWClass.hpp"
+#include "API/CNWCCMessageData.hpp"
+#include "API/CNWRules.hpp"
+#include "API/CNWSCreature.hpp"
+#include "API/CNWSCreatureStats.hpp"
+#include "API/CNWSItem.hpp"
+#include "API/CNWSpellArray.hpp"
+#include "API/CNWSPlayer.hpp"
+#include "API/Constants.hpp"
+#include "API/Functions.hpp"
+#include "API/Globals.hpp"
+#include "API/CAppManager.hpp"
+#include "API/CServerExoApp.hpp"
+#include "API/CTwoDimArrays.hpp"
+#include "API/C2DA.hpp"
+
+namespace Tweaks {
+
+using namespace NWNXLib;
+using namespace NWNXLib::API;
+
+FixScrollLearningBug::FixScrollLearningBug(Services::HooksProxy* hooker)
+{
+    hooker->RequestExclusiveHook<Functions::_ZN12CNWSCreature11LearnScrollEj>
+                                    (&CNWSCreature__LearnScroll_hook);
+}
+
+BOOL FixScrollLearningBug::CNWSCreature__LearnScroll_hook(CNWSCreature *pThis, OBJECT_ID oidScrollToLearn)
+{
+
+    auto pServerExoApp = Globals::AppManager()->m_pServerExoApp;
+
+    auto pItem = pServerExoApp->GetItemByGameObjectID(oidScrollToLearn);
+    if (!pItem || pItem->m_nBaseItem != Constants::BaseItem::SpellScroll) {
+        pThis->SendFeedbackMessage(0x4e, nullptr, nullptr);
+        return false;
+    }
+
+    auto pActiveProperty = pItem->GetActiveProperty(0);
+    if (!pActiveProperty)
+        return false;
+
+    int nSpellIndex;
+    Globals::Rules()->m_p2DArrays->m_pIPRPSpells->GetINTEntry(pActiveProperty->m_nSubType, CExoString("SpellIndex"), &nSpellIndex);
+    
+    auto pSpell = Globals::Rules()->m_pSpellArray->GetSpell(nSpellIndex);
+    if (!pSpell)
+        return false;
+
+    auto pPlayer = pServerExoApp->GetClientObjectByObjectId(pThis->m_idSelf);
+    if (pPlayer && !pPlayer->HasExpansionPack(Globals::Rules()->GetSpellExpansionLevel(nSpellIndex), 1))
+        return false;
+
+    CItemRepository* pItemRepository = nullptr;
+    auto pPossessor = pServerExoApp->GetGameObject(pItem->m_oidPossessor);
+    if (pPossessor) {
+        if (pPossessor->m_nObjectType == Constants::ObjectType::Creature) {
+            pItemRepository = pPossessor->AsNWSCreature()->m_pcItemRepository;
+        } else if (pPossessor->m_nObjectType == Constants::ObjectType::Item) {
+            pItemRepository = pPossessor->AsNWSItem()->m_pItemRepository;
+        }
+    }
+
+    if (!pItemRepository) //Not present in decompiled code
+        return 0;
```

Comment:
Here you suddenly switch from returning `false` to `0`. `true`/`false` is much clearer.